lib/src/widgets/board_editor.dart

```diff
@@ -0,0 +1,152 @@
+import 'package:chessground/src/fen.dart';
+import 'package:flutter/widgets.dart';
+
+import 'piece.dart';
+import 'positioned_square.dart';
+import '../models.dart';
+import '../board_editor_settings.dart';
+
+/// A chessboard widget where pieces can be dragged around freely (including dragging piece off and onto the board).
+///
+/// This widget can be used as the basis for a fully fledged board editor, similar to https://lichess.org/editor.
+class BoardEditor extends StatelessWidget {
+  const BoardEditor({
+    super.key,
+    required this.size,
+    required this.orientation,
+    required this.fen,
+    this.settings = const BoardEditorSettings(),
+    this.onTappedSquare,
+    this.onDroppedPiece,
+    this.onDiscardedPiece,
+  });
+
+  /// Visual size of the board.
+  final double size;
+
+  double get squareSize => size / 8;
+
+  /// The FEN of the position to display.
+  final String fen;
+
+  /// Settings that control the appearance of the board editor.
+  final BoardEditorSettings settings;
+
+  /// Side by which the board is oriented.
+  final Side orientation;
+
+  /// Called when the given [square] was tapped.
+  final void Function(SquareId square)? onTappedSquare;
+
+  /// Called when a [piece] has been dragged to a new [destination] square.
+  /// If [origin] is not `null`, the piece was dragged from that square of the board editor.
+  /// Otherwise, it was dragged from outside the board editor.
+  /// Each square of the board is a DragTarget<Piece>, so to drop your own piece widgets
+  /// onto the board, put them in a Draggable<Piece> and set the data to the piece you want to drop.
+  final void Function(SquareId? origin, SquareId destination, Piece piece)?
+      onDroppedPiece;
+
+  /// Called when a piece that was originally at the given [square] was dragged off the board.
+  final void Function(SquareId square)? onDiscardedPiece;
+
+  @override
+  Widget build(BuildContext context) {
+    SquareId? draggedPieceOrigin;
```

Comment:
I don't think we should keep track of that as a local variable of `build`.

In that case we must convert `BoardEditor` to a `StatefulWidget` and make `draggedPieceOrigin` a property of the State.